pose_prediction/plugin.cpp

```diff
@@ -18,12 +19,12 @@ class pose_prediction_impl : public pose_prediction {
 
 		// Make the first valid fast pose be straight ahead.
 		if (first_time && pose_ptr) {
-			std::lock_guard<std::recursive_mutex> lock {offset_mutex};
+			auto pose = correct_pose(*pose_ptr);
+			std::unique_lock lock {offset_mutex};
 			// check again, now that we have mutual exclusion
 			if (first_time) {
-				auto pose = correct_pose(*pose_ptr);
-				const_cast<pose_prediction_impl&>(*this).set_offset(pose.orientation);
-				const_cast<pose_prediction_impl&>(*this).first_time = false;
+				first_time = false;
+				offset = pose.orientation;
```

Comment:
This diff is not functionally equivalent. Originally, `offset` was being updated in `set_offset` with some matrix math. Now, it just gets the value of the current pose and `set_offset` is no longer called. Why is that?

Reply:
`offset` is the Identity matrix to begin with, so multiplying by the inverse is a no-op.

Reply:
I can't call `set_offset` because the lock is no longer recursive (can't acquire it, and then call a method that acquires it).

I need to acquire it here, to lock out other "first time calls" to `pose_predict`; The lock picks exactly one call which "wins the race" to be the "first time call"

Reply:
The locking is fine. That is not my concern. My concern is that `set_offset` returns `pose.orientation.inverse` the very first time because `offset` is the identity quaternion to begin with. The above line returns just `post.orientation`, without the inverse. What am I missing here?

Reply:
I think it should be inverted, so I changed this in the implementation.

Failing to invert the initial pose may give approximately correct behavior, if the initial pose was a quaternion, $q = q^{-1}$. There are many such quaternions, e.g. if the angular displacement is close to 0 or close to 180, on any plane.